### PR TITLE
Convert FileOperationsHandler + FileTagsHandler to helpers

### PR DESCRIPTION
**Pull request by itsWindows11**
**Resolved / Related Issues**
Items resolved / related issues by this PR.
- Related #4180.
- Related #9858

**Validation**
How did you test these changes?
- [x] Built and ran the app
- [ ] Tested the changes for accessibility

**Screenshots (optional)**
Add screenshots here.

### CODE REVIEW

File: src/Files.App/DataModels/RemoteDataObject.cs [hunk 1]
```diff
@@ -0,0 +1,300 @@
...
+using System.Runtime.InteropServices;
+using System.Runtime.InteropServices.ComTypes;
+using System.Runtime.Versioning;
+using System.Windows.Forms;
+using Vanara.PInvoke;
+using STATSTG = System.Runtime.InteropServices.ComTypes.STATSTG;
+
+namespace Files.App.DataModels
+{
+    // Class taken from Rx-Explorer (https://github.com/zhuxb711/RX-Explorer)
```
Comment (yaichenbaum) [inline src/Files.App/DataModels/RemoteDataObject.cs:14]: We are we taking someone else's code?

Comment (itsWindows11) [reply src/Files.App/DataModels/RemoteDataObject.cs:14]: I copied the file from the one in Files.FullTrust so I wasn't the one who wrote this comment.

File: src/Files.App/Helpers/UIFilesystemHelpers.cs [hunk 2]
```diff
@@ -101,18 +101,11 @@ public static async void CutItem(IShellPage associatedInstance)
                         if (connection != null)
                         {
                             string filePaths = string.Join('|', associatedInstance.SlimContentPage.SelectedItems.Select(x => x.ItemPath));
```
Comment (hez2010) [inline src/Files.App/Helpers/UIFilesystemHelpers.cs:103]: Given that the interop between Files.App and Files.FullTrust is no longer needed, we shouldn't serialize paths to a string anymore. Instead we should pass the `string[]` for paths directly.

File: src/Files.App/Filesystem/FilesystemOperations/ShellFilesystemOperations.cs [hunk 3]
```diff
@@ -98,37 +98,19 @@ public async Task<IStorageHistory> CopyItemsAsync(IList<IStorageItemWithPath> so
...
-                    { "filepath", string.Join('|', sourceRename.Select(s => s.Path)) },
-                    { "destpath", string.Join('|', destinationRename) },
-                    { "overwrite", false },
-                    { "HWND", NativeWinApiHelper.CoreWindowHandle.ToInt64() }
-                });
-                result &= (FilesystemResult)(status == AppServiceResponseStatus.Success
-                    && response.Get("Success", defaultJson).GetBoolean());
-                var shellOpResult = JsonSerializer.Deserialize<ShellOperationResult>(response.Get("Result", defaultJson).GetString());
-                copyResult.Items.AddRange(shellOpResult?.Final ?? Enumerable.Empty<ShellOperationItemResult>());
+                var resultItem = await FileOperationsHelpers.CopyItemAsync(string.Join('|', sourceRename.Select(s => s.Path)), string.Join('|', destinationRename), false, NativeWinApiHelper.CoreWindowHandle.ToInt64(), operationID);
```
Comment (hez2010) [inline src/Files.App/Filesystem/FilesystemOperations/ShellFilesystemOperations.cs:101]: Given that the interop between Files.App and Files.FullTrust is no longer needed, we shouldn't serialize paths to a string anymore. Instead we should pass the `string[]` for paths directly.

File: src/Files.App/Filesystem/FilesystemOperations/ShellFilesystemOperations.cs [hunk 4]
```diff
@@ -98,37 +98,19 @@ public async Task<IStorageHistory> CopyItemsAsync(IList<IStorageItemWithPath> so
...
-                    { "filepath", string.Join('|', sourceReplace.Select(s => s.Path)) },
-                    { "destpath", string.Join('|', destinationReplace) },
-                    { "overwrite", true },
-                    { "HWND", NativeWinApiHelper.CoreWindowHandle.ToInt64() }
-                });
-                result &= (FilesystemResult)(status == AppServiceResponseStatus.Success
-                    && response.Get("Success", defaultJson).GetBoolean());
-                var shellOpResult = JsonSerializer.Deserialize<ShellOperationResult>(response.Get("Result", defaultJson).GetString());
-                copyResult.Items.AddRange(shellOpResult?.Final ?? Enumerable.Empty<ShellOperationItemResult>());
+                var resultItem = await FileOperationsHelpers.CopyItemAsync(string.Join('|', sourceReplace.Select(s => s.Path)), string.Join('|', destinationReplace), true, NativeWinApiHelper.CoreWindowHandle.ToInt64(), operationID);
```
Comment (hez2010) [inline src/Files.App/Filesystem/FilesystemOperations/ShellFilesystemOperations.cs:109]: Given that the interop between Files.App and Files.FullTrust is no longer needed, we shouldn't serialize paths to a string anymore. Instead we should pass the `string[]` for paths directly.

File: src/Files.App/Filesystem/FilesystemOperations/ShellFilesystemOperations.cs [hunk 5]
```diff
@@ -395,19 +350,10 @@ public async Task<IStorageHistory> DeleteItemsAsync(IList<IStorageItemWithPath>
...
-                { "fileop", "DeleteItem" },
-                { "operationID", operationID },
-                { "filepath", string.Join('|', deleleFilePaths) },
-                { "permanently", permanently },
-                { "HWND", NativeWinApiHelper.CoreWindowHandle.ToInt64() }
-            });
-            var result = (FilesystemResult)(status == AppServiceResponseStatus.Success
-                && response.Get("Success", defaultJson).GetBoolean());
-            var shellOpResult = JsonSerializer.Deserialize<ShellOperationResult>(response.Get("Result", defaultJson).GetString());
+            var (success, deleteResult) = await FileOperationsHelpers.DeleteItemAsync(string.Join('|', deleleFilePaths), permanently, NativeWinApiHelper.CoreWindowHandle.ToInt64(), operationID);
```
Comment (hez2010) [inline src/Files.App/Filesystem/FilesystemOperations/ShellFilesystemOperations.cs:353]: Given that the interop between Files.App and Files.FullTrust is no longer needed, we shouldn't serialize paths to a string anymore. Instead we should pass the `string[]` for paths directly.

File: src/Files.App/Filesystem/FilesystemOperations/ShellFilesystemOperations.cs [hunk 6]
```diff
@@ -525,37 +471,17 @@ public async Task<IStorageHistory> MoveItemsAsync(IList<IStorageItemWithPath> so
...
-                    { "filepath", string.Join('|', sourceRename.Select(s => s.Path)) },
-                    { "destpath", string.Join('|', destinationRename) },
-                    { "overwrite", false },
-                    { "HWND", NativeWinApiHelper.CoreWindowHandle.ToInt64() }
-                });
-                result &= (FilesystemResult)(status == AppServiceResponseStatus.Success
-                    && response.Get("Success", defaultJson).GetBoolean());
-                var shellOpResult = JsonSerializer.Deserialize<ShellOperationResult>(response.Get("Result", defaultJson).GetString());
-                moveResult.Items.AddRange(shellOpResult?.Final ?? Enumerable.Empty<ShellOperationItemResult>());
+                var (status, response) = await FileOperationsHelpers.MoveItemAsync(string.Join('|', sourceRename.Select(s => s.Path)), string.Join('|', destinationRename), false, NativeWinApiHelper.CoreWindowHandle.ToInt64(), operationID);
```
Comment (hez2010) [inline src/Files.App/Filesystem/FilesystemOperations/ShellFilesystemOperations.cs:474]: Given that the interop between Files.App and Files.FullTrust is no longer needed, we shouldn't serialize paths to a string anymore. Instead we should pass the `string[]` for paths directly.

File: src/Files.App/Filesystem/FilesystemOperations/ShellFilesystemOperations.cs [hunk 7]
```diff
@@ -525,37 +471,17 @@ public async Task<IStorageHistory> MoveItemsAsync(IList<IStorageItemWithPath> so
...
-                    { "filepath", string.Join('|', sourceReplace.Select(s => s.Path)) },
-                    { "destpath", string.Join('|', destinationReplace) },
-                    { "overwrite", true },
-                    { "HWND", NativeWinApiHelper.CoreWindowHandle.ToInt64() }
-                });
-                result &= (FilesystemResult)(status == AppServiceResponseStatus.Success
-                    && response.Get("Success", defaultJson).GetBoolean());
-                var shellOpResult = JsonSerializer.Deserialize<ShellOperationResult>(response.Get("Result", defaultJson).GetString());
-                moveResult.Items.AddRange(shellOpResult?.Final ?? Enumerable.Empty<ShellOperationItemResult>());
+                var (status, response) = await FileOperationsHelpers.MoveItemAsync(string.Join('|', sourceReplace.Select(s => s.Path)), string.Join('|', destinationReplace), true, NativeWinApiHelper.CoreWindowHandle.ToInt64(), operationID);
```
Comment (hez2010) [inline src/Files.App/Filesystem/FilesystemOperations/ShellFilesystemOperations.cs:481]: Given that the interop between Files.App and Files.FullTrust is no longer needed, we shouldn't serialize paths to a string anymore. Instead we should pass the `string[]` for paths directly.

File: src/Files.App/Filesystem/FilesystemOperations/ShellFilesystemOperations.cs [hunk 8]
```diff
@@ -764,20 +682,10 @@ public async Task<IStorageHistory> RestoreItemsFromTrashAsync(IList<IStorageItem
...
-                { "filepath", string.Join('|', source.Select(s => s.Path)) },
-                { "destpath", string.Join('|', destination) },
-                { "overwrite", false },
-                { "HWND", NativeWinApiHelper.CoreWindowHandle.ToInt64() }
-            });
-            var result = (FilesystemResult)(status == AppServiceResponseStatus.Success
-                && response.Get("Success", defaultJson).GetBoolean());
-            var shellOpResult = JsonSerializer.Deserialize<ShellOperationResult>(response.Get("Result", defaultJson).GetString());
-            moveResult.Items.AddRange(shellOpResult?.Final ?? Enumerable.Empty<ShellOperationItemResult>());
+            var (status, response) = await FileOperationsHelpers.MoveItemAsync(string.Join('|', source.Select(s => s.Path)), string.Join('|', destination), false, NativeWinApiHelper.CoreWindowHandle.ToInt64(), operationID);
```
Comment (hez2010) [inline src/Files.App/Filesystem/FilesystemOperations/ShellFilesystemOperations.cs:685]: Given that the interop between Files.App and Files.FullTrust is no longer needed, we shouldn't serialize paths to a string anymore. Instead we should pass the `string[]` for paths directly.

File: src/Files.App/Filesystem/FilesystemOperations/ShellFilesystemOperations.cs [hunk 9]
```diff
@@ -935,24 +832,8 @@ private async Task<DialogResult> GetFileListDialog(IEnumerable<string> source, s
...
-                });
-                if (status == AppServiceResponseStatus.Success && response.ContainsKey("Processes"))
-                {
-                    return JsonSerializer.Deserialize<List<Win32Process>>(response["Processes"]);
-                }
-            }
-            return null;
-        }
+        private List<Win32Process> WhoIsLocking(IEnumerable<string> filesToCheck)
+            => (List<Win32Process>?)FileOperationsHelpers.CheckFileInUse(string.Join('|', filesToCheck)) ?? new List<Win32Process>();
```
Comment (hez2010) [inline src/Files.App/Filesystem/FilesystemOperations/ShellFilesystemOperations.cs:836]: Given that the interop between Files.App and Files.FullTrust is no longer needed, we shouldn't serialize paths to a string anymore. Instead we should pass the `string[]` for paths directly.

File: src/Files.App/Helpers/FileOperationsHelpers.cs [hunk 10]
```diff
@@ -0,0 +1,918 @@
...
+            return targetHandle.ToInt64();
+        }
+
+        public static Task SetClipboard(string fileToCopy, DataPackageOperation operation)
+        {
+            return Win32API.StartSTATask(() =>
+            {
+                System.Windows.Forms.Clipboard.Clear();
+                var fileList = new System.Collections.Specialized.StringCollection();
+                fileList.AddRange(fileToCopy.Split('|'));
```
Comment (hez2010) [inline src/Files.App/Helpers/FileOperationsHelpers.cs:56]: Change this to use `string[]` directly.

File: src/Files.App/Helpers/FileOperationsHelpers.cs [hunk 11]
```diff
@@ -0,0 +1,918 @@
...
+                    deleteTcs.TrySetResult(false);
+                }
+
+                return (await deleteTcs.Task, shellOperationResult);
+            });
+        }
+
+        public static Task<(bool, ShellOperationResult)> DeleteItemAsync(string filePath, bool permanently, long ownerHwnd, string operationID = "")
+        {
+            var fileToDeletePath = filePath.Split('|');
```
Comment (hez2010) [inline src/Files.App/Helpers/FileOperationsHelpers.cs:244]: Change this to use `string[]` directly.

File: src/Files.App/Helpers/FileOperationsHelpers.cs [hunk 12]
```diff
@@ -0,0 +1,918 @@
...
+                progressHandler.RemoveOperation(operationID);
+
+                return (await renameTcs.Task, shellOperationResult);
+            });
+        }
+
+        public static Task<(bool, ShellOperationResult)> MoveItemAsync(string filePath, string destPath, bool overwriteOnMove, long ownerHwnd, string operationID = "")
+        {
+            var fileToMovePath = filePath.Split('|');
+            var moveDestination = destPath.Split('|');
```
Comment (hez2010) [inline src/Files.App/Helpers/FileOperationsHelpers.cs:388]: Change these to use `string[]` directly.

File: src/Files.App/Helpers/UIFilesystemHelpers.cs [hunk 13]
```diff
@@ -207,18 +200,11 @@ public static async Task CopyItem(IShellPage associatedInstance)
...
```
Comment (hez2010) [inline src/Files.App/Helpers/UIFilesystemHelpers.cs:202]: Given that the interop between Files.App and Files.FullTrust is no longer needed, we shouldn't serialize paths to a string anymore. Instead we should pass the `string[]` for paths directly.

File: src/Files.App/Helpers/FileOperationsHelpers.cs [hunk 14]
```diff
@@ -0,0 +1,918 @@
...
+                    }, App.Logger);
+                }
+
+                return (await createTcs.Task, shellOperationResult);
+            });
+        }
+
+        public static Task<(bool, ShellOperationResult)> TestRecycleAsync(string filePath)
+        {
+            var fileToDeletePath = filePath.Split('|');
```
Comment (hez2010) [inline src/Files.App/Helpers/FileOperationsHelpers.cs:172]: Change this to use `string[]` directly.

File: src/Files.App/Helpers/FileOperationsHelpers.cs [hunk 15]
```diff
@@ -0,0 +1,918 @@
...
+                progressHandler.RemoveOperation(operationID);
+
+                return (await moveTcs.Task, shellOperationResult);
+            });
+        }
+
+        public static Task<(bool, ShellOperationResult)> CopyItemAsync(string filePath, string destPath, bool overwriteOnCopy, long ownerHwnd, string operationID = "")
+        {
+            var fileToCopyPath = filePath.Split('|');
+            var copyDestination = destPath.Split('|');
```
Comment (hez2010) [inline src/Files.App/Helpers/FileOperationsHelpers.cs:465]: Change these to use `string[]` directly.

File: src/Files.App/Helpers/FileOperationsHelpers.cs [hunk 16]
```diff
@@ -0,0 +1,918 @@
...
+                return (await copyTcs.Task, shellOperationResult);
+            });
+        }
+
+        public static void TryCancelOperation(string operationId)
+            => progressHandler.TryCancel(operationId);
+
+        public static IEnumerable<Win32Process>? CheckFileInUse(string filePath)
+        {
+            var fileToCheckPath = filePath.Split('|');
```
Comment (hez2010) [inline src/Files.App/Helpers/FileOperationsHelpers.cs:545]: Change this to use `string[]` directly.

**Review by hez2010 — CHANGES_REQUESTED** [review]
Did a rough review and left some suggestions.

**Comment by itsWindows11** [comment]
hmm
```
---> System.IO.IOException: The process cannot access the file 'C:\Users\itswi\AppData\Local\Packages\FilesDev_ykqwq8d6ps0ag\LocalState\filetags.db' because it is being used by another process.
```

**Comment by hez2010** [comment]
Can you verify is Files still able to launch? I cannot launch Files on my computer with this PR anymore, it hangs in the background constantly.

**Comment by itsWindows11** [comment]
> Can you verify is Files still able to launch? I cannot launch Files on my computer with this PR anymore, it hangs in the background constantly. 

Works for me, dunno why it shouldn't work for you

**Comment by hez2010** [comment]
> Works for me, dunno why it shouldn't work for you

Did you try deploying a Release package and run it?

**Comment by itsWindows11** [comment]
Nope, will try in a few hours

**Comment by hez2010** [comment]
May related: https://github.com/files-community/Files/pull/10031
Will test again after that PR merged.

File: src/Files.App/Filesystem/FilesystemOperations/ShellFilesystemOperations.cs [hunk 17]
```diff
@@ -240,38 +222,23 @@ public async Task<(IStorageHistory, IStorageItem)> CreateAsync(IStorageItemWithP
...
-                            (status, response) = await connection.SendMessageForResponseAsync(new ValueSet()
-                            {
-                                { "Arguments", "FileOperation" },
-                                { "fileop", "CreateFile" },
-                                { "filepath", source.Path },
-                                { "template", newEntryInfo?.Template },
-                                { "data", newEntryInfo?.Data }
-                            });
-                        }
+                            (success, response) = await FileOperationsHelpers.CreateItemAsync(source.Path, "CreateFile", newEntryInfo?.Template, Convert.ToBase64String(newEntryInfo?.Data));
```
Comment (hez2010) [inline src/Files.App/Filesystem/FilesystemOperations/ShellFilesystemOperations.cs:230]: I don't think the `Convert.ToBase64String` is necessary. We can pass the `byte[]` directly without serialization.

File: src/Files.App/Helpers/FileOperationsHelpers.cs [hunk 18]
```diff
@@ -0,0 +1,909 @@
...
+                catch
+                {
+                    createTcs.TrySetResult(false);
+                }
+
+                if (dataStr != null && (shellOperationResult.Items.SingleOrDefault()?.Succeeded ?? false))
+                {
+                    SafetyExtensions.IgnoreExceptions(() =>
+                    {
+                        var dataBytes = Convert.FromBase64String(dataStr);
```
Comment (hez2010) [inline src/Files.App/Helpers/FileOperationsHelpers.cs:159]: Same as above.

File: src/Files.App/Helpers/FileOperationsHelpers.cs [hunk 19]
```diff
@@ -0,0 +1,909 @@
...
+using Vanara.Windows.Shell;
+using Windows.ApplicationModel.DataTransfer;
+using Windows.Foundation.Collections;
+
+namespace Files.App.Helpers
+{
+    public class FileOperationsHelpers
+    {
+        private static FileTagsDb dbInstance = new(Path.Combine(Windows.Storage.ApplicationData.Current.LocalFolder.Path, "filetags.db"));
+        private static readonly JsonElement defaultJson = JsonSerializer.SerializeToElement("{}");
```
Comment (hez2010) [inline src/Files.App/Helpers/FileOperationsHelpers.cs:32]: Is this still necessary?

**Review by hez2010 — CHANGES_REQUESTED** [review]

**Comment by hez2010** [comment]
> May related: https://github.com/files-community/Files/pull/10031
> Will test again after that PR merged.

Can confirm it's #10031.

Other findings:
- Paste: doesn't work
- Delete: doesn't work
- Property-Security: doesn't work
- Create items: doesn't work

File: src/Files.App/Helpers/FileOperationsHelpers.cs (hunk 19 above)
Comment (itsWindows11) [reply src/Files.App/Helpers/FileOperationsHelpers.cs:32]: For the DB instance yes, I'll work on removing the JSON default later

**Comment by hez2010** [comment]
> hmm, I don't think there's another process that's using this file
> 
> ```
> ---> System.IO.IOException: The process cannot access the file 'C:\Users\itswi\AppData\Local\Packages\FilesDev_ykqwq8d6ps0ag\LocalState\filetags.db' because it is being used by another process.
> ```

This is expected becuase now we are trying to open filetags.db across multiple Files instances, which is not supported by LiteDB. You can leave this problem as is and open a new issue for it. We need to use something like mutex to make sure there's only one connection simultaneously.

**Comment by itsWindows11** [comment]
> Other findings:
> - Paste: doesn't work
> - Delete: doesn't work
> - Property-Security: doesn't work
> - Create items: doesn't work

Is that related to #10031 too?

**Comment by hez2010** [comment]
Properties - Details doesn't work.

**Comment by itsWindows11** [comment]
> Properties - Details doesn't work.

I can reproduce this even when building from the main branch so doesn't seem related to this PR.

> When paste or delete multiple files, the "ongoing task" doesn't reflect the process properly.
> Paste and Delete still doesn't work if there're multiple Files instances running simultaneously.

Will take a look.

**Comment by itsWindows11** [comment]
Any news? cc. @hez2010 @yair100

**Comment by hez2010** [comment]
Sorry for the delay. Will take care of this PR in next few days.

**Comment by yaichenbaum** [comment]
@itsWindows11 almost ready to look at this, can you resolve the merge conflicts?

**Comment by itsWindows11** [comment]
Sure thing, expect some commits tomorrow.

**Review by yair100 — APPROVED** [review]
~~LGTM~~
Just noticed the build error

**Comment by itsWindows11** [comment]
Doesn't look like a build error I can fix; I can compile the app on my end.

**Review by hez2010 — APPROVED** [review]
LGTM and tested ok except when it comes to  multiple files instances. 
@yaichenbaum I think it's fine to merge it as is. 
We will need to change how we use LiteDB to open the connection and close it immediately after finishing db operations, and use a Mutex for inter-process synchronization. I'm currently working on this and will open a follow-up PR.

**Comment by yaichenbaum** [comment]
@itsWindows11 thank you for all you work and keeping up with this PR!

**Pull request merged**